Disable polymorphisation

davidtwco opened the pull request:
Fixes #74614.

This PR disables polymorphisation to fix the regression in #74614 after investigation into the issue makes it clear that the fix won't be trivial. ~~I'll file an issue shortly to replace #74614 with the findings so far.~~ #74636 has been filed to track the fix of the underlying regression.

r? @eddyb

davidtwco committed:
mir: improve polymorphic constant bug

This commit changes the span and content of the "collection encountered
polymorphic constant" bug in monomorphization collection to point to the
use of the constant rather than the definition.

Signed-off-by: David Wood <david@davidtw.co>

davidtwco committed:
sess: disable polymorphisation

This commit disables polymorphisation to resolve regressions related to
closures which inherit unused generic parameters and are then used in
casts or reflection.

Signed-off-by: David Wood <david@davidtw.co>

lcnr commented on src/librustc_mir/monomorphize/collector.rs:
```diff
@@ -627,8 +627,9 @@ impl<'a, 'tcx> MirVisitor<'tcx> for MirNeighborCollector<'a, 'tcx> {
-                        "collection encountered polymorphic constant",
+                        self.body.source_info(location).span,
+                        "collection encountered polymorphic constant: {}",
+                        constant
```
```suggestion
                        substituted_constant
```

davidtwco replied:
Fixed!

lcnr reviewed (approved):
Can you add the following as a test?
```rust
fn test<T>() { std::mem::size_of::<T>(); }

pub fn foo<T>(_: T) -> &'static fn() {
    &(test::<T> as fn())
}

fn outer<T>() {
    foo(|| ());
}


fn main() {
    outer::<u8>();
}
```

wesleywiser commented:
r=me with the test added. We shouldn't roll this up since it will swing perf results again.

@bors rollup=never

davidtwco committed:
tests: add regression test for #74614

This commit adds a regression test for #74614 so that it is fixed before
polymorphisation is re-enabled.

Signed-off-by: David Wood <david@davidtw.co>

wesleywiser commented:
@bors r+ p=1 

Raising the priority because of nightly breakage

bors commented:
:pushpin: Commit 799d52eff7e96fc339327966b056ab89c061b255 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 799d52eff7e96fc339327966b056ab89c061b255 -->
<!-- homu: {"type":"Approved","sha":"799d52eff7e96fc339327966b056ab89c061b255","approver":"wesleywiser"} -->

bors commented:
:hourglass: Testing commit 799d52eff7e96fc339327966b056ab89c061b255 with merge 16b5082861b7528b4f3d171bd6797961b906fdd5...
<!-- homu: {"type":"BuildStarted","head_sha":"799d52eff7e96fc339327966b056ab89c061b255","merge_sha":"16b5082861b7528b4f3d171bd6797961b906fdd5"} -->

Manishearth commented:
@bors retry yield

yielding to rollup

Aaron1011 commented:
@bors retry

lqd commented:
@bors p=10 for nightly breakage

bors commented:
:hourglass: Testing commit 799d52eff7e96fc339327966b056ab89c061b255 with merge bbebe7351fcd29af1eb9a35e315369b15887ea09...
<!-- homu: {"type":"BuildStarted","head_sha":"799d52eff7e96fc339327966b056ab89c061b255","merge_sha":"bbebe7351fcd29af1eb9a35e315369b15887ea09"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/900216026), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35502)
Approved by: wesleywiser
Pushing bbebe7351fcd29af1eb9a35e315369b15887ea09 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35502","checks-actions":"https://github.com/rust-lang-ci/rust/runs/900216026"},"merge_sha":"bbebe7351fcd29af1eb9a35e315369b15887ea09"} -->

nnethercote commented:
#69749 was a [perf loss](https://perf.rust-lang.org/compare.html?start=734233d29771869f824d8ddbaddabb90b3b68e03&end=b52522ade1f6979a35b24087dadcf5ba5c981cbe&stat=instructions:u), but disabling polymorphization was [mostly neutral](https://perf.rust-lang.org/compare.html?start=9e92106d457abd14f82adc29e7f2496861e07916&end=bbebe7351fcd29af1eb9a35e315369b15887ea09&stat=instructions:u). @davidtwco: is that expected?

Mark-Simulacrum commented:
@davidtwco, @wesleywiser -- could someone follow up on this being a performance-neutral PR? Thanks!

davidtwco commented:
> #69749 was a [perf loss](https://perf.rust-lang.org/compare.html?start=734233d29771869f824d8ddbaddabb90b3b68e03&end=b52522ade1f6979a35b24087dadcf5ba5c981cbe&stat=instructions:u), but disabling polymorphization was [mostly neutral](https://perf.rust-lang.org/compare.html?start=9e92106d457abd14f82adc29e7f2496861e07916&end=bbebe7351fcd29af1eb9a35e315369b15887ea09&stat=instructions:u). @davidtwco: is that expected?

> @davidtwco, @wesleywiser -- could someone follow up on this being a performance-neutral PR? Thanks!

@njn @Mark-Simulacrum Apologies for the delay in looking into this - this was surprising but I think it makes sense; I did some local benchmarking of the PR which landed polymorphisation (with `-Zpolymorphisation=off`) and the merge before that.

I believe that the majority of the losses that we didn't win back are from cross-crate metadata/incr-comp changes - `-Zpolymorphize=off` makes `unused_generic_params` always return every-parameter-is-used, but that result is still serialized - there was nothing to serialize without polymorphisation having landed. Analysing perf results isn't something I've got much experience with, but as far as I can tell, this explains the inconsistency here.

Mark-Simulacrum commented:
Thanks! Is that something we could plausibly "fix", or does that seem too hard? I guess we're already encoding them as bitsets. Maybe we can skip recording all-zero bitsets and decode them appropriately as well? It may also be worth switching to a lower minimum bound -- u64 implies 64 generics which I expect is way above the normal amount for most functions?

davidtwco commented:
> Thanks! Is that something we could plausibly "fix", or does that seem too hard? 

I think both of these suggestions make sense, I've implemented them in #75155.
